### ISSUE TITLE
 how can I use the standard c++ STL library ?

### ISSUE THREAD

**Issue by patriquetrique** [Bigfoot71/raymob#21]
hi, I'm sure the solution is obvious but I'm really missing the point.

when I try to include the headers of the STL library (to use everything that belongs to the std: namespace: you know what I'm talking about) in my main.cpp the compilation does not find the linked files and the header is highlighted as an error.
It’s almost as if the main.cpp was compiled as a .c by force.

This is my first time with grdle and android studio so I imagine that there is a modification to be made in the settings but after trying several solutions idea I think that I just don't understand how it works.

Thanks in advance if anyone has the patience to answer me.

**Comment by patriquetrique** [comment]
I searched a little more and I reinstalled and reconfigured Android Studio thinking that it might be a configuration error.

the same problem persists and I can't find any explanation even from colleagues more experienced than me.

I think the problem comes from a subtlety specific to raymob that I haven't understood.

the problem is that the mention of any include line of headers from C++ is highlighted as an error and when I run the debug of the main.cpp I get some compilation error for each C++ headers

I tried to isolate them to see if the problem comes from a compatibility problem with only some but none of them are working, the following example is with iostream but I get the same error with <string>, <unordered_map>, <thread> and others all native to C++

the compilation errors in question :

'iostream' file not found [x86_64] :

C:\Users\AppData\Local\Android\Sdk\ndk\26.1.10909125\toolchains\llvm\prebuilt\windows-x86_64\bin\clang++.exe --target=x86_64-none-linux-android24 --sysroot=C:/Users/AppData/Local/Android/Sdk/ndk/26.1.10909125/toolchains/llvm/prebuilt/windows-x86_64/sysroot -DDEBUG -DPLATFORM_ANDROID -D_DEBUG -Draymob_EXPORTS -IC:/Users/AppData/Local/Android/Sdk/ndk/26.1.10909125/sources/android/native_app_glue -IC:/Users/OneDrive/Desktop/code/raymob-master/app/src/main/cpp/deps/raylib -IC:/Users/OneDrive/Desktop/code/raymob-master/app/src/main/cpp/deps/raymob -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security   -g -fno-limit-debug-info -fPIC -MD -MT CMakeFiles/raymob.dir/main.cpp.o -MF CMakeFiles\raymob.dir\main.cpp.o.d -o CMakeFiles/raymob.dir/main.cpp.o -c C:/Users/OneDrive/Desktop/code/raymob-master/app/src/main/cpp/main.cpp

C:/Users/OneDrive/Desktop/code/raymob-master/app/src/main/cpp/main.cpp:3:10: fatal error: 'iostream' file not found
#include <iostream>
         ^~~~~~~~~~

and

com.android.ide.common.process.ProcessException: ninja: Entering directory `C:\Users\kevin\OneDrive\Desktop\code\raymob-master\app\.cxx\Debug\e44705hf\x86_64' :

this one is very long, for a question of readability I could give it to you in a separate message if you need more detail.

thank you for paying attention to my request

**Comment by patriquetrique** [comment]
I searched a little more and I reinstalled and reconfigured Android Studio thinking that it might be a configuration error.

the same problem persists and I can't find any explanation even from colleagues more experienced than me.

I think the problem comes from a subtlety specific to raymob that I haven't understood.

the problem is that the mention of any include line of headers from C++ is highlighted as an error and when I run the debug of the main.cpp I get some compilation error for each C++ headers

I tried to isolate them to see if the problem comes from a compatibility problem with only some but none of them are working, the following example is with iostream but I get the same error with , <unordered_map>, and others all native to C++

the compilation errors in question :

'iostream' file not found [x86_64] :

C:\Users\AppData\Local\Android\Sdk\ndk\26.1.10909125\toolchains\llvm\prebuilt\windows-x86_64\bin\clang++.exe --target=x86_64-none-linux-android24 --sysroot=C:/Users/AppData/Local/Android/Sdk/ndk/26.1.10909125/toolchains/llvm/prebuilt/windows-x86_64/sysroot -DDEBUG -DPLATFORM_ANDROID -D_DEBUG -Draymob_EXPORTS -IC:/Users/AppData/Local/Android/Sdk/ndk/26.1.10909125/sources/android/native_app_glue -IC:/Users/OneDrive/Desktop/code/raymob-master/app/src/main/cpp/deps/raylib -IC:/Users/OneDrive/Desktop/code/raymob-master/app/src/main/cpp/deps/raymob -g -DANDROID -fdata-sections -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -g -fno-limit-debug-info -fPIC -MD -MT CMakeFiles/raymob.dir/main.cpp.o -MF CMakeFiles\raymob.dir\main.cpp.o.d -o CMakeFiles/raymob.dir/main.cpp.o -c C:/Users/OneDrive/Desktop/code/raymob-master/app/src/main/cpp/main.cpp

C:/Users/OneDrive/Desktop/code/raymob-master/app/src/main/cpp/main.cpp:3:10: fatal error: 'iostream' file not found
#include
^~~~~~~~~~

and

com.android.ide.common.process.ProcessException: ninja: Entering directory `C:\Users\OneDrive\Desktop\code\raymob-master\app.cxx\Debug\e44705hf\x86_64' :

this one is very long, for a question of readability I could give it to you in a separate message if you need more detail.

thank you for paying attention to my request

**Comment by Bigfoot71** [comment]
@patriquetrique Apologies for the delay; I was unavailable over the past few days.

This should work without requiring any additional actions on your part, I just tested it on my end without any issues.

Since you mentioned that you've reconfigured everything cleanly, what happens if you delete the following directories and then rebuild/relaunch the project?  
- `[raymob]/app/.cxx`  
- `[raymob]/app/build`  

(where `[raymob]` is the root of your project or the location where you cloned the repository).

**Comment by Bigfoot71** [comment]
The current `CMakeLists.txt` collects all C/C++ source files in the project while excluding the `deps` directory, which is intended to contain submodules [[L18](https://github.com/Bigfoot71/raymob/blob/master/app/src/main/cpp/CMakeLists.txt#L18)]:

```cmake
# Fetch all source files for your project (recursively), excluding 'deps' source files
file(GLOB_RECURSE SOURCES "${CMAKE_SOURCE_DIR}/*.c" "${CMAKE_SOURCE_DIR}/*.cpp")
list(FILTER SOURCES EXCLUDE REGEX "${CMAKE_SOURCE_DIR}/deps/.*")
```

**By default, you don’t need to manually add `main.cpp` to the `SOURCES`.**  

This `CMakeLists.txt` isn't the cleanest solution, but it’s probably the easiest way to quickly set up and test a project without worrying too much about CMake in the beginning.  

**You have two options:**  
1. Revert to the default `CMakeLists.txt`.  
2. Remove the two lines I highlighted above that automatically gather all sources while excluding `deps`, and instead manually specify your source files one by one (which is the proper way to do it).

**Then, delete the `build` and `.cxx` directories.**  

The `.cxx` directory is a build cache and can cause issues when source files are enumerated automatically as in the default `CMakeLists.txt`.  

**Afterward, try rebuilding the project and let me know what happens. There shouldn’t be any issues.**  

If the problem persists, you can share an example of your project where you’re encountering these errors, and I’ll check it on my side.

**Comment by Bigfoot71** [comment]
> whether the lines are present or not, the compilation did not change, it worked without the STL headers but not with them and presented the same errors.

Wait, can you build an example that defines a class, for instance, without including any STL headers?  
If that’s the case, there’s likely an issue with your setup.  

Feel free to send me a reproducible example—you should be able to share a zip archive here.

**Comment by Bigfoot71** [comment]
Errors occur when decompressing your zip file on my end. I’ve tried several tools without success. It appears to be corrupted:
```
$ unzip raymob-master-bouncing_ball.zip
Archive:  raymob-master-bouncing_ball.zip
  End-of-central-directory signature not found.  Either this file is not
  a zipfile, or it constitutes one disk of a multi-part archive. 
```
___

I don’t think this will solve your problem, it was mainly a suggestion. AFAIK, CMake should omit duplicates anyway.

To manually specify the sources (which is recommended), delete these lines:  
```cmake
# Fetch all source files for your project (recursively), excluding 'deps' source files
file(GLOB_RECURSE SOURCES "${CMAKE_SOURCE_DIR}/*.c" "${CMAKE_SOURCE_DIR}/*.cpp")
list(FILTER SOURCES EXCLUDE REGEX "${CMAKE_SOURCE_DIR}/deps/.*")
```

And replace them with the following:  
```cmake
set(SOURCES ${CMAKE_SOURCE_DIR}/main.cpp)
```

You can define as many sources as you like, like this:  
```cmake
set(SOURCES
    ${CMAKE_SOURCE_DIR}/source1.cpp
    ${CMAKE_SOURCE_DIR}/source2.cpp
    ${CMAKE_SOURCE_DIR}/source3.cpp
)
```

If you have already defined your `SOURCES` variable (or another variable), you can include other elements as you did before:  
```cmake
set(SOURCES ${CMAKE_SOURCE_DIR}/main.cpp)
list(APPEND SOURCES ${CMAKE_SOURCE_DIR}/other.cpp)
```

**Comment by Bigfoot71** [comment]
Yes, it's very strange. I finally managed to unzip your file on Windows with 7-Zip. I tested it on my work PC, I only uncommented the `#include <string>` in the main file, and everything built successfully. Here's even a screenshot of the app, which confirms that there is an issue on your side.

![Screenshot_20241126_204557](https://github.com/user-attachments/assets/d3a2de5d-57f3-4140-a917-fcad79d4a264)

You can still share the complete logs, preferably in a text file or a pastebin, however you'd like, and I can try to understand where the issue is coming from. But if you say your device often malfunctions and the errors change without having made any modification, it doesn’t sound very good :/

But it might just be a wrongly specified path somewhere, you never know.

**Comment by patriquetrique** [comment]
thank you very much for your help and time !
you were right, I spent the evening tweaking my configs, my path and my installations.
I formatted my computer and reinstalled android studio and the SDK and reconfigured the linked paths and updated the NDK to the latest version.

I don't really know if the problem came from a system obsolescence or a typo error in my configuration but everything works correctly now.

thank you again for your help and thank you for raymob !
